### PR TITLE
add language extensions: SAFE and UNSAFE

### PR DESCRIPTION
**Pull request by smunix**
Hobbes functions such as `unsafeCast` are, as labelled, deeply unsafe. They escape the type-checker safety guarantees and can crash hi server if applied from a remote client to the wrong parameter type.

This change introduces support for language extensions, among which the `UNSAFE` pragma that can be used to flag expression variables that won't escape the scope of the server they are being run into. By doing this, they prevent remote client from invoking them, unless otherwise stated.

Also, the current fix introduces a new mode, `Safe`, under which the `UNSAFE` extension is effective. `SAFE` is meant to override this feature, and can only be run from the server side.

### CODE REVIEW

File: include/hobbes/lang/module.H [hunk 1]
```diff
@@ -164,6 +188,8 @@ template <typename T>
     virtual T with(const MVarDef*)     const = 0;
     virtual T with(const ClassDef*)    const = 0;
     virtual T with(const InstanceDef*) const = 0;
+    virtual T with(const MUnsafePragmaDef*)  const = 0;
```
Comment (kthielen) [inline include/hobbes/lang/module.H:191]: Doesn't this seem a little overly specific for this level of definition?  If you wanted to support custom pragmas (don't we already have something for this?) you wouldn't want to add a new type for each one here right?

File: lib/hobbes/boot/amapping.hob [hunk 2]
```diff
@@ -52,7 +52,9 @@ instance (p++s=a, Function f p rf, Function rf s r) => Function f a r where
 class Array xs x | xs -> x where
   size     :: xs -> long
   element  :: (xs, long) -> x
+  elementM  :: (xs, long) -> (()+x)
   elements :: (xs, long, long) -> [x]
+  elementsM :: (xs, long, long) -> [x]
```
Comment (kthielen) [inline lib/hobbes/boot/amapping.hob:57]: Why is `elements` in on this?  Isn't it always safe?

File: lib/hobbes/boot/farray.hob [hunk 3]
```diff
@@ -21,10 +23,12 @@ anyi f i xs =
     true
   else
     (anyi(f,i+1L,xs))
+{-# UNSAFE anyi #-}
 
 // [doc] determines whether any element of the array satisfied the predicate
 any :: (a -> bool, [a]) -> bool
 any f xs = anyi(f,0L,xs)
+{-# SAFE any #-}
```
Comment (kthielen) [inline lib/hobbes/boot/farray.hob:31]: No way, every single function?  Whose idea was this and why is this needed?

**Comment by kthielen** [comment]
This looks like a massive change with little added value.  We're really going to start annotating every single function as "SAFE" versus "UNSAFE"?  Where did this request come from?  Does this serve a real need?

**Review by kthielen — CHANGES_REQUESTED** [review]
Let's not do this.

File: lib/hobbes/boot/farray.hob (hunk 3 above)
Comment (smunix) [reply lib/hobbes/boot/farray.hob:31]: Yes, every single unsafe core function is annotated accordingly in this PR.
```sh
> anyi(\_.false,100l,[0..10])
fish: “hi” terminated by signal SIGSEGV (Address boundary error)
```
`any`, however, is safe

**Review by smunix — COMMENTED** [review]
Availability and robustness are required for our applications. Not every single function needs to be annotated, only the ones that system admins have flagged unsafe.

File: lib/hobbes/boot/amapping.hob (hunk 2 above)
Comment (kthielen) [reply lib/hobbes/boot/amapping.hob:57]: Isn't that just a bug?  There's nothing about the type that excludes a safe result, unlike `element`.  Why build ceremony/scaffolding around a bug?

File: lib/hobbes/boot/farray.hob (hunk 3 above)
Comment (kthielen) [reply lib/hobbes/boot/farray.hob:31]: I don't think that anyone is actually using `anyi`, but if they do then why not just fix the indexing bug?  What's the point of adding all of these annotations? What's the point of adding a `SAFE` annotation on `any`?  Who actually asked for this?

**Comment by kthielen** [comment]
> Availability and robustness are required for our applications.

That's fine, if not a little vague, but why would that mean adding annotation and ceremony everywhere, rather than fixing an indexing bug or changing a type signature?

Ultimately, the simpler this thing is, the easier it is to verify its safety and the easier it is to maintain.  I would be on board with personally helping to make it simpler also toward that end of making it more robust.  But if we're going to go the way of ever increasing entropy, then we should just drop it now.

**Comment by smunix** [comment]
The alternative, changing type signatures, was found to be highly intrusive and would push costly breaking changes down to existing users. This is still under evaluation though.

Some  expressions (e.g. the family of unsafeCast, newPrim functions, ...) can't be made safe due to their very "unsafe" nature, or because they're C-defined functions that are exported by Hobbes.

Also, some users write "unsafe" helper functions that are more performant than the similar ones that implement some heavy validation. These users are required to either expose the "unsafe" versions or not, or do provide both validating and not validating versions of expressions, depending on their configuration.

These are some of the cases this PR is addressing. The case of existing bugs is orthogonal to the issues above, and is being addressed separately.

File: lib/hobbes/boot/amapping.hob (hunk 2 above)
Comment (smunix) [reply lib/hobbes/boot/amapping.hob:57]: will be addressed in #373

**Comment by kthielen** [comment]
I guess these findings and user expectations aren't part of the public record, so we can't evaluate alternative approaches or consider whether new requirements are consistent with old ones.  It's hard to imagine a process like that working in the long run, for an open source project at least.

I guess at some point there will be concerns beyond memory safety, and another class of "SAFE" annotations will have to be added for access to the file system, or other OS resources that might be sensitive.  At some point it'll be another type system, but distinct from the regular type system, and simplifying it will be found to be highly intrusive, so it'll keep growing that way.

I guess this is the sort of situation that Guy Steele had in mind with his "growing a language" essay, and already it's too late to reverse other expedient decisions.

File: lib/hobbes/boot/amapping.hob [hunk 4]
```diff
@@ -157,8 +157,9 @@ instance Array (vector t s n) t where
 
 reverse :: [a] -> [a]
 reverse xs = selectElementsInto(xs, element, 0L, size(xs), newArray(size(xs)), lsub(size(xs),1L), lneg(1L))
+{-# SAFE reverse #-}
 
-// sometimes it's a trivial/obvious single-step to array-ness
+// sometimes it's a trivial/obvious single-step to array-nessß
```
Comment (kthielen) [inline lib/hobbes/boot/amapping.hob:162]: `ß`?

Comment (smunix) [reply lib/hobbes/boot/amapping.hob]: fixed

**Review by chenkai036 — APPROVED** [review]

**Comment by kthielen** [comment]
I'm sorry that it's come to this, and I know that the lack of support for this project over the years has forced expedient changes to get rammed through like this, I don't blame you guys.

It's always been a fight against inertia and skepticism, even when the project had significant unambiguous improvements over existing infrastructure.  I will always say that I was able to prove with hobbes that a generic structurally typed variant of Haskell can be effectively used in latency-sensitive systems as well as for storing and querying large and complex datasets in ways that traditional tick databases can't.  There have been some bumps in the road, but this project has also done new things that haven't been done before.  Despite the mean reversion here, eventually this model is going to win out.

Good luck.  I will work on a fork or new project, and will keep folks I've met who've been following this project in the loop.
